src/main/java/org/testng/TestNG.java

```diff
@@ -1394,7 +1394,9 @@ private SuiteRunner createSuiteRunner(XmlSuite xmlSuite) {
     }
 
     for (IReporter r : result.getReporters()) {
-      addListener(r);
+      if (!m_reporters.containsKey(r.getClass())) {
```

Comment:
Did you consider to replace the line by `maybeAddListener(m_reporters, r.getClass(), r);`?